omegaconf/base.py

```diff
@@ -396,120 +382,198 @@ def _select_impl(
+            parent=root,
             key=last_key,
             value=value,
             throw_on_missing=throw_on_missing,
             throw_on_resolution_failure=throw_on_resolution_failure,
         )
         return root, last_key, value
 
-    def _resolve_simple_interpolation(
+    def _resolve_interpolation_from_parse_tree(
         self,
+        parent: Optional["Container"],
+        value: "Node",
         key: Any,
-        inter_type: str,
-        inter_key: str,
+        parse_tree: OmegaConfGrammarParser.ConfigValueContext,
         throw_on_missing: bool,
         throw_on_resolution_failure: bool,
     ) -> Optional["Node"]:
-        from omegaconf import OmegaConf
+        from .nodes import StringNode
 
-        from .nodes import ValueNode
+        resolved = self.resolve_parse_tree(
+            parse_tree=parse_tree,
+            key=key,
+            parent=parent,
+            throw_on_missing=throw_on_missing,
+            throw_on_resolution_failure=throw_on_resolution_failure,
+        )
 
-        inter_type = ("str:" if inter_type is None else inter_type)[0:-1]
-        if inter_type == "str":
-            root_node, inter_key = self._resolve_key_and_root(inter_key)
-            parent, last_key, value = root_node._select_impl(
-                inter_key,
-                throw_on_missing=throw_on_missing,
-                throw_on_resolution_failure=throw_on_resolution_failure,
+        if resolved is None:
+            return None
+        elif isinstance(resolved, str):
+            # Result is a string: create a new StringNode for it.
+            return StringNode(
+                value=resolved,
+                key=key,
+                parent=parent,
+                is_optional=value._metadata.optional,
             )
+        else:
+            assert isinstance(resolved, Node)
+            return resolved
+
+    def _resolve_node_interpolation(
+        self,
+        inter_key: str,
+        throw_on_missing: bool,
+        throw_on_resolution_failure: bool,
+    ) -> Optional["Node"]:
+        """A node interpolation is of the form `${foo.bar}`"""
+        root_node, inter_key = self._resolve_key_and_root(inter_key)
+        parent, last_key, value = root_node._select_impl(
+            inter_key,
+            throw_on_missing=throw_on_missing,
+            throw_on_resolution_failure=throw_on_resolution_failure,
+        )
+
+        if parent is None or value is None:
+            if throw_on_resolution_failure:
+                raise ConfigKeyError(f"Interpolation key '{inter_key}' not found")
+            else:
+                return None
+        assert isinstance(value, Node)
+        return value
 
-            # if parent is None or (value is None and last_key not in parent):  # type: ignore
-            if parent is None or value is None:
+    def _evaluate_custom_resolver(
+        self,
+        key: Any,
+        # parent: Optional["Container"],
```

Comment:
left over?

Reply:
> left over?

Right, trivially removed in 310f22b => resolving